File: app/src/main/java/com/github/ljarka/bakingapp/widget/WidgetConfigurationActivity.java
package com.github.ljarka.bakingapp.widget;

import android.app.Activity;
import android.appwidget.AppWidgetManager;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Intent;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.widget.Toast;

import com.github.ljarka.bakingapp.MainViewModel;
import com.github.ljarka.bakingapp.OnRecipeItemClickListener;
import com.github.ljarka.bakingapp.R;
import com.github.ljarka.bakingapp.RecipesAdapter;
import com.github.ljarka.bakingapp.network.model.Recipe;

import java.util.List;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.disposables.Disposables;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;

public class WidgetConfigurationActivity extends AppCompatActivity implements OnRecipeItemClickListener {
    private RecipesAdapter adapter = new RecipesAdapter();
    private int appWidgetId;
    private Disposable disposable = Disposables.empty();

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_widget_configuration);
        appWidgetId = getIntent().getExtras().getInt(AppWidgetManager.EXTRA_APPWIDGET_ID, AppWidgetManager.INVALID_APPWIDGET_ID);

        MainViewModel viewModel = ViewModelProviders.of(this).get(MainViewModel.class);
        RecyclerView recyclerView = findViewById(R.id.recyclerView);
        recyclerView.setLayoutManager(new GridLayoutManager(this, getResources().getInteger(R.integer.span_count)));
        recyclerView.setAdapter(adapter);
        adapter.setOnRecipeItemClickListener(this);
        disposable = viewModel.getRecipes()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Consumer<List<Recipe>>() {
                    @Override
                    public void accept(List<Recipe> recipes) throws Exception {
                        adapter.setRecipes(recipes);
                    }
                }, new Consumer<Throwable>() {
                    @Override
                    public void accept(Throwable throwable) throws Exception {
                        Toast.makeText(WidgetConfigurationActivity.this, "Error", Toast.LENGTH_LONG).show();
                    }
                });
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        disposable.dispose();
    }

    @Override
    public void onRecipeItemClick(@Nullable Recipe recipe) {
        PreferenceManager.getDefaultSharedPreferences(this)
                .edit()
                .putString(String.valueOf(appWidgetId), recipe.getName())
                .apply();
        IngredientsWidgetProvider.updateWidget(this, recipe.getName(), recipe.getIngredients(),
                AppWidgetManager.getInstance(this), appWidgetId);
        Intent resultValue = new Intent();
        resultValue.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId);
        setResult(Activity.RESULT_OK, resultValue);
        finish();
    }
}

File: app/src/main/java/com/github/ljarka/bakingapp/widget/WidgetConfigurationActivity.kt
package com.github.ljarka.bakingapp.widget

import android.app.Activity
import android.app.Activity.RESULT_OK
import android.appwidget.AppWidgetManager
import android.arch.lifecycle.ViewModelProviders
import android.content.Intent
import android.os.Bundle
import android.preference.PreferenceManager
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.GridLayoutManager
import android.widget.Toast
import com.github.ljarka.bakingapp.MainViewModel
import com.github.ljarka.bakingapp.OnRecipeItemClickListener
import com.github.ljarka.bakingapp.R
import com.github.ljarka.bakingapp.RecipesAdapter
import com.github.ljarka.bakingapp.network.model.Recipe
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.activity_main.*

class WidgetConfigurationActivity : AppCompatActivity(), OnRecipeItemClickListener {
    private val adapter: RecipesAdapter = RecipesAdapter()
    private var appWidgetId: Int? = null
    private lateinit var viewModel: MainViewModel
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_widget_configuration)
        appWidgetId = intent.extras.getInt(
                AppWidgetManager.EXTRA_APPWIDGET_ID,
                AppWidgetManager.INVALID_APPWIDGET_ID)

        viewModel = ViewModelProviders.of(this).get(MainViewModel::class.java)
        recyclerView.layoutManager = GridLayoutManager(this, resources.getInteger(R.integer.span_count))
        recyclerView.adapter = adapter
        adapter.onRecipeItemClickListener = this
        viewModel.getRecipes()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    adapter.recipes = it
                }, {
                    Toast.makeText(this@WidgetConfigurationActivity, "Error", Toast.LENGTH_LONG).show()
                })
    }

    override fun onRecipeItemClick(recipe: Recipe?) {
        PreferenceManager
                .getDefaultSharedPreferences(this)
                .edit().putString(appWidgetId.toString(), recipe?.name)
                .apply()
        IngredientsWidgetProvider.updateWidget(this, recipe?.name, recipe?.ingredients, AppWidgetManager.getInstance(this),
                appWidgetId)
        val resultValue = Intent()
        resultValue.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId)
        setResult(Activity.RESULT_OK, resultValue)
        finish()
    }
}